### PR TITLE
Improve handling of invisible sequences in trim mode

### PR DESCRIPTION
**Pull request by stroncium**
Added tests from #32 and fixed code to pass them.

### CODE REVIEW

**Comment by sindresorhus** [comment]
I merged https://github.com/chalk/wrap-ansi/pull/32, can you rebase and remove the `.failing` part from that test instead? I think it makes more sense to keep it as a separate test.

**Comment by sindresorhus** [comment]
// @coreyfarrell

**Comment by coreyfarrell** [comment]
@sindresorhus I've just realized another question - should strings be wrapped before or after trimming?  Currently wrapping is done before trimming which produces the following results:

```js
// 3 spaces before / between / after
t.is(wrapAnsi('   foo   bar   ', 6), 'foo\nbar');
t.is(wrapAnsi('   foo   bar   ', 3), '\nfoo\nbar\n');
```

**Comment by stroncium** [comment]
@coreyfarrell @sindresorhus It's not a question of trim first or not(trimming is not just a stage), it's a bug, and I guess part of it was introduced by my big patch. Anyhow, added test, fixed.

**Comment by coreyfarrell** [comment]
@stroncium I just verified that all `cliui` tests now pass when `wrap-ansi` code is updated with this PR.  Thanks!

File: index.js [hunk 1]
```diff
@@ -60,6 +60,30 @@ const wrapWord = (rows, word, columns) => {
...
+const stringVisibleTrimSpaces = str => {
+	const words = str.split(' ');
+	let first = 0;
+	let last = words.length;
+	while (first < last) {
+		if (stringWidth(words[first]) > 0) {
+			break;
+		}
+
+		first++;
```
Comment (coreyfarrell) [inline index.js]: The tests never hit this line.  Is it possible for `words.length > 0 && stringWidth(words[0]) === 0`?

Comment (stroncium) [reply index.js]: Yeah, in actual code left side is always trimmed already. Changed to trimRight

File: index.js [hunk 2]
```diff
@@ -60,6 +60,23 @@ const wrapWord = (rows, word, columns) => {
...
+
+	while (last > first) {
+		if (stringWidth(words[last - 1]) > 0) {
+			break;
+		}
+
+		last--;
+	}
+
+	return words.slice(0, first).join('') + words.slice(first, last).join(' ') + words.slice(last, words.length).join('');
```
Comment (coreyfarrell) [inline index.js]: Please remove `words.slice(0, first).join('')` as it is always an empty string.

Personally I'd also remove the `first` variable entirely and just use `0` where appropriate though I don't feel strongly about this.

Comment (coreyfarrell) [reply index.js]: @stroncium ping

@sindresorhus do you see any other issues with this PR?

Comment (stroncium) [reply index.js]: @coreyfarrell done with a bonus. Sorry for taking so long. I suppose github decided I've read this when I glanced over email. =/ (Particularly annoying feature, if anyone knows how to disable this I'd be happy to listen.)

Comment (coreyfarrell) [reply index.js]: Github includes a link to a 1x1 pixel image like ``.  Maybe if you've told your email client that github emails are trusted this bypasses the normal privacy protection?  Every github email shows me the warning `To protect your privacy, Thunderbird has blocked remote content in this message.`

Comment (stroncium) [reply index.js]: Well, we'll see. I'm using Inbox, but it gets shut down today, hopefully gmail got this feature.

**Comment by sindresorhus** [comment]
Thanks for fixing, @stroncium 👍

**Pull request merged**